lib/philomena/users.ex

```diff
@@ -943,7 +943,10 @@ defmodule Philomena.Users do
   defp setup_roles(nil), do: nil
 
   defp setup_roles(user) do
-    role_map = Map.new(user.roles, &{&1.resource_type || &1.name, &1.name})
+    role_map =
+      user.roles
+      |> Enum.group_by(& &1.resource_type, & &1.name)
+      |> Map.new(fn {type, names} -> {type, Map.new(names, &{&1, []})} end)
```

Comment:
This empty list dummy value in the map looks a bit weird, could use a `MapsSet` instead, but I get that this way it's easier to pattern-match in canada checks.

Maybe instead of using a `Map` here we could just keep a scalar role name here by finding the maximum from the role names. For example `["moderator", "admin"]` should coerce to just `"admin"` and `["moderator"]` should coerce to just `"moderator"`. Could use an `Enum.max_by` for that. This way we won't need any change in `ability.ex`.

Reply:
I thought of this and I don't think this is a fair assumption for this system

Reply:
The permission model is a little scuffed because some places bypass Canada and depend on user's role. And "roles" are more capabilities than roles, yes. But if it works it works. We don't really need complicated access control.